practice5.java

```diff
@@ -0,0 +1,94 @@
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+// 1. Реализуйте структуру телефонной книги с помощью HashMap, учитывая, что 1 
+// человек может иметь несколько телефонов.
+// 2. Пусть дан список сотрудников:Иван Иванов, Светлана Петрова, Кристина Белова, 
+// Анна Мусина, Анна Крутова, Иван Юрин, Петр Лыков, Павел Чернов, Петр Чернышов, 
+// Мария Федорова, Марина Светлова, Мария Савина, Мария Рыкова, Марина Лугова, 
+// Анна Владимирова, Иван Мечников, Петр Петин, Иван Ежов. Написать программу, 
+// которая найдет и выведет повторяющиеся имена с количеством повторений. 
+// Отсортировать по убыванию популярности. Для сортировки можно использовать 
+// TreeMap, мы не успели, но я расписал пример использования и прикрепил к 
+// материалам урока в файле TestTreeMap
+// 3. *Реализовать алгоритм пирамидальной сортировки (HeapSort)
+
+public class practice5 {
+    public static void main(String[] args) {
+        // 1
+        TelBook tb = new TelBook();
+        tb.add("Вася", "+125874693254", "+258741963123");
+        System.out.println(tb.toString());
+
+        // 2
+        String[] str = new String[] { "Иван Иванов", "Светлана Петрова", "Кристина Белова",
+                "Анна Мусина", "Анна Крутова", "Иван Юрин", "Петр Лыков", "Павел Чернов", "Петр Чернышов",
+                "Мария Федорова", "Марина Светлова", "Мария Савина", "Мария Рыкова", "Марина Лугова",
+                "Анна Владимирова", "Иван Мечников", "Петр Петин", "Иван Ежов" };
+        System.out.println(getSorted(str));
+        // 3
+        int arr[] = new int[] { 5, 10, 7, 2, 11, 1, 25, 6 };
+        int n = arr.length;
+        System.out.println("Исходный массив: " + Arrays.toString(arr));
+        heapSort(arr, n);
+        System.out.println("Отсортированный массив: " + Arrays.toString(arr));
+    }
+
+    static SortedSet<Map.Entry<String, Integer>> getSorted(String[] str) {
+        Map<String, Integer> map = new TreeMap<String, Integer>();
+        for (int i = 0; i < str.length - 2; i++) {
+            int count = 1;
+            if (!map.containsKey(str[i].split(" ")[0])) {
+                for (int j = i + 1; j < str.length; j++) {
+                    if (str[i].split(" ")[0].equals(str[j].split(" ")[0])) {
+                        count++;
+                        map.put(str[i].split(" ")[0], count);
+                    }
+                }
+            }
+        }
+
+        SortedSet<Map.Entry<String, Integer>> sorted = new TreeSet<Map.Entry<String, Integer>>(
+                Comparator.<Map.Entry<String, Integer>, Integer>comparing(Map.Entry::getValue)
+                        .reversed()
+                        .thenComparing(Map.Entry::getKey));
+
+        sorted.addAll(map.entrySet());
+        return sorted;
+    }
+

```

Comment:
static Map<Integer, ArrayList<String>> ex(String[] s){
        Map<Integer, ArrayList<String>> map = new TreeMap<>(Comparator.reverseOrder());
        Set<String> names = new HashSet<>();
        for (String name: s){
            String firstName = name.split(" ")[0];
            if (!names.contains(firstName)){
                names.add(firstName);
                int count = 0;
                for (String fullName: s){
                    if (fullName.split(" ")[0].equals(firstName)){
                        count++;
                    }
                }
                if (map.containsKey(count)){
                    ArrayList list = map.get(count);
                    list.add(firstName);
                } else {
                    ArrayList<String> list = new ArrayList<>();
                    list.add(firstName);
                    map.put(count, list);
                }
            }
        }
        return map;
    }